org.mapstruct.eclipse/src/org/mapstruct/eclipse/internal/quickfix/MapStructQuickFix.java

```diff
@@ -0,0 +1,205 @@
+
+import java.util.Iterator;
+
+import org.eclipse.core.resources.IMarker;
+import org.eclipse.core.resources.IResource;
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.jdt.core.ICompilationUnit;
+import org.eclipse.jdt.core.IJavaElement;
+import org.eclipse.jdt.core.JavaCore;
+import org.eclipse.jdt.core.dom.AST;
+import org.eclipse.jdt.core.dom.ASTNode;
+import org.eclipse.jdt.core.dom.CompilationUnit;
+import org.eclipse.jdt.core.dom.ImportDeclaration;
+import org.eclipse.jdt.core.dom.rewrite.ASTRewrite;
+import org.eclipse.jdt.internal.ui.JavaPlugin;
+import org.eclipse.jdt.internal.ui.JavaPluginImages;
+import org.eclipse.jdt.internal.ui.javaeditor.EditorUtility;
+import org.eclipse.jdt.internal.ui.text.correction.ASTResolving;
+import org.eclipse.jdt.ui.SharedASTProvider;
+import org.eclipse.jface.text.Position;
+import org.eclipse.jface.text.source.Annotation;
+import org.eclipse.jface.text.source.IAnnotationModel;
+import org.eclipse.swt.graphics.Image;
+import org.eclipse.ui.IEditorInput;
+import org.eclipse.ui.IMarkerResolution2;
+import org.eclipse.ui.texteditor.MarkerAnnotation;
+import org.mapstruct.eclipse.internal.quickfix.visitors.FindMethodByPositionVisitor;
+
+/**
+ * Base class for quick fixes
+ *
+ * @author Andreas Gudian
+ */
+@SuppressWarnings("restriction")
+public abstract class MapStructQuickFix implements IMarkerResolution2 {
+    private String label;
+
+    protected MapStructQuickFix(String label) {
+        this.label = label;
+    }
+
+    @Override
+    public String getLabel() {
+        return label;
+    }
+
+    @Override
+    public void run(IMarker marker) {
+        ICompilationUnit compilationUnit = null;
+        try {
+            IResource resource = marker.getResource();
+            IJavaElement javaElement = JavaCore.create( resource );
+
+            compilationUnit = (ICompilationUnit) javaElement.getAdapter( ICompilationUnit.class );
+            IEditorInput input = EditorUtility.getEditorInput( compilationUnit );
+            if ( input != null ) {
+                CompilationUnit astCompilationUnit = toAST( compilationUnit );
+
+                ASTNode locatedNode =
+                    locateASTNodeForSartingOffset( findProblemStart( input, marker ), astCompilationUnit );
+                if ( locatedNode != null ) {
+                    ASTRewrite rewrite = getASTRewrite( astCompilationUnit, locatedNode, marker );
+
+                    if ( rewrite != null ) {
+                        compilationUnit.applyTextEdit( rewrite.rewriteAST(), null );
+                        compilationUnit.becomeWorkingCopy( null );
+                        compilationUnit.commitWorkingCopy( true, null );
+                        compilationUnit.discardWorkingCopy();
+                        marker.delete();
+                    }
+                }
+            }
+        }
+        catch ( CoreException e ) {
+            e.printStackTrace();
```

Comment:
Tried it out: the user isn't bothered with the error - it just finds its way into the error-log. So I'll rethrow it here.